### PR TITLE
PHYDF-14 API catalog, PHYDF-15 API grade

### PR DESCRIPTION
**Pull request by UvgenGen**
[PHYDF-14](https://youtrack.raccoongang.com/issue/PHYDF-14) API catalog
[PHYDF-15](https://youtrack.raccoongang.com/issue/PHYDF-15) API grade

### CODE REVIEW

File: lms/djangoapps/course_api/serializers.py [hunk 1]
```diff
@@ -77,6 +84,16 @@ class CourseSerializer(serializers.Serializer):  # pylint: disable=abstract-meth
...
 
+    def get_skilltag(self, course_overview):
+        """
+        Get the representation for SerializerMethodField `skilltag`
+        """
+        skilltag = course_overview.skilltag
+        try:
+            return json.loads(skilltag)
+        except ValueError:
+            return "Error. Data is not value"
```
Comment (cmltaWt0) [inline lms/djangoapps/course_api/serializers.py]: Typo. "Error. Data is not valid"

File: common/lib/xmodule/xmodule/course_module.py [hunk 2]
```diff
@@ -167,6 +167,15 @@ def to_json(self, values):
 
 
 class CourseFields(object):
+    main_topic = String(display_name='1Main topic', scope=Scope.settings,
+                                 help=_('free text (can only be one value)'),
```
Comment (cmltaWt0) [inline common/lib/xmodule/xmodule/course_module.py]: Please arrange according to the OpenEdx python style guide.

File: lms/djangoapps/grades/api/views.py [hunk 3]
```diff
@@ -149,8 +161,15 @@ def get(self, request, course_id):
 
         prep_course_for_grading(course, request)
         course_grade = CourseGradeFactory().create(request.user, course)
-
+        user = User.objects.get(username=username)
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/api/views.py]: Handle non existent user please.

File: lms/djangoapps/grades/api/views.py [hunk 4]
```diff
@@ -149,8 +161,15 @@ def get(self, request, course_id):
...
+        course_enrollment = CourseEnrollment.objects.get(course_id=CourseKey.from_string(course_id), user=user.id)
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/api/views.py]: THe same as above - if the User is not exists we will have an exception.

File: lms/djangoapps/grades/api/views.py [hunk 5]
```diff
@@ -149,8 +161,15 @@ def get(self, request, course_id):
...
+        grade = PersistentCourseGrade.objects.get(user_id=user.id, course_id=CourseKey.from_string(course_id))
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/api/views.py]: THe same as above.
Also pleae handle incorrect `course_id`.

File: lms/djangoapps/grades/api_grade.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+class APIGrade(object):
+    """
+    Send user data to the Gamification server
+    """
+    def __init__(self):
+        self.is_enabled = settings.FEATURES.get('ENABLE_API_GRADE', False)
+        if self.is_enabled:
+            self.API_GRADE_PROPERTIES = settings.FEATURES.get('API_GRADE_PROPERTIES', {})
+            if not self.GAMMA_PROPERTIES:
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/api_grade.py]: Change for `API_GRADE_PROPERTIES `

File: lms/djangoapps/grades/api_grade.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            "main_topic": main_topic,
+            "skilltag": skilltag,
+            "course_level": course_level,
+            "effort": effort
+        }
+        headers = {
+            'x-functions-key': self.API_GRADE_PROPERTIES['APP_KEY'],
+        }
+        requests.post(
+            '{}/orchestrators/feedbackloop/{}}/?code={}'.format(self.API_GRADE_PROPERTIES['API_URL'], userId, self.API_GRADE_PROPERTIES['APP_SECRET']),
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/api_grade.py]: Long string - add indentation please.

File: lms/djangoapps/grades/signals/handlers.py [hunk 8]
```diff
@@ -199,6 +202,28 @@ def recalculate_course_grade(sender, course, course_structure, user, **kwargs):
     Updates a saved course grade.
     """
     CourseGradeFactory().update(user, course, course_structure)
+    grade = CourseGradeFactory()._get_saved_grade(user, course, course_structure)
+    if grade._percent == 1.0:
+        course_enrollment = CourseEnrollment.objects.get(course_id=course.id, user=user.id)
+        persist_course_grade = PersistentCourseGrade.objects.get(user_id=user.id, course_id=course.id)
+        duration = persist_course_grade.passed_timestamp - course_enrollment.created if persist_course_grade.passed_timestamp else None
+        import pdb; pdb.set_trace()
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/signals/handlers.py]: Remove debugger.

File: lms/djangoapps/grades/signals/handlers.py [hunk 9]
```diff
@@ -199,6 +202,28 @@ def recalculate_course_grade(sender, course, course_structure, user, **kwargs):
...
+        data = {
+            "contentProvider": "FastLane",
+            "userId": user.id,
+            "courseId": course.id.to_deprecated_string(),
+            "lastlogin": user.last_login,
+            "percentageOfcompletion": grade.percent,
+            "duration": duration,
+            "lastVisit": '---',
+            "completationDate" : persist_course_grade.passed_timestamp,
```
Comment (cmltaWt0) [inline lms/djangoapps/grades/signals/handlers.py]: Can it be retrieved from `grade` not from `persist_course_grade ` ?

Comment (UvgenGen) [reply lms/djangoapps/grades/signals/handlers.py]: `grade` does not contain an attribute `passed_timestamp`

**Pull request merged**